Changing Bastion instance spec doesn't work

/kind bug

**What steps did you take and what happened:**
* Deploy a cluster with a bastion
* Once the cluster is deployed, edit OpenStackCluster.Spec to disable the bastion like it's [documented](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/5cc483bfc6eae8a8b8a67b32e9b7af0bafa473ca/api/v1alpha8/openstackcluster_types.go#L157-L159). So later we can re-create it with another spec (e.g. new flavor).
* Removing `OpenStackCluster.Spec.Bastion` produces an error in CAPO:

```
I0202 02:37:59.557716      20 controller.go:115] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="default/development-10884-rm62s" namespace="default" name="development-10884-rm62s" reconcileID="e0ebc724-7686-42ee-80fc-ce9984c1beb5"
```
(Here CAPO crashed in my lab, I had to restart it).

* Ok this time let's just switch `enabled` to False:
The bastion is removed and re-created immediately!
And `enabled` is True again (by itself).


**What did you expect to happen:**
The bastion should be deleted, so we can re-create it with a new spec.


**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): current main
- Logs from when we delete the bastion to when it's re-created: https://hastebin.com/share/sabasopulu.python

I got a similar (maybe the same) error today when I tried to remove a bastion host. This is the last 2 logs of the capo controller pod.
```
I0206 08:51:48.033845       1 openstackcluster_controller.go:312] "Reconciling Bastion" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" co
ntrollerKind="OpenStackCluster" OpenStackCluster="capi-cluster/viktor-capi-elastx-sc" namespace="capi-cluster" name="viktor-capi-elastx-sc" reconcileID="6cbf2347-5dca-4de
9-8646-6f902a45bd3a" cluster="viktor-capi-elastx-sc"                                                                                                                      
I0206 08:51:48.582046       1 controller.go:115] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="openstackc
luster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="capi-cluster/viktor-capi-elastx-sc" namespace="capi-cluster"
 name="viktor-capi-elastx-sc" reconcileID="6cbf2347-5dca-4de9-8646-6f902a45bd3a"                                                                                          
panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                                                       
        panic: runtime error: invalid memory address or nil pointer dereference                                                                                           
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1c0b8a4]                                                                                                  
                                                                                                                                                                          
goroutine 376 [running]:                                                                                                                                                  
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()                                                                                    
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:116 +0x1fa                                                               
panic({0x1e1a4e0, 0x35bd790})                                                                                                                                             
        /usr/local/go/src/runtime/panic.go:884 +0x213                                                                                                                     
sigs.k8s.io/cluster-api-provider-openstack/controllers.bastionToInstanceSpec(0xc000681000, {0xc0005e6840?, 0xc000681000?})                                                
        /workspace/controllers/openstackcluster_controller.go:393 +0xc4                                                                                                   
sigs.k8s.io/cluster-api-provider-openstack/controllers.deleteBastion({0x2466e58?, 0xc0011f5040?}, 0xc000cb0ea0, 0xc000681000)                                             
        /workspace/controllers/openstackcluster_controller.go:240 +0x3c5                                                                                                  
sigs.k8s.io/cluster-api-provider-openstack/controllers.reconcileBastion({0x2466e58, 0xc0011f5040}, 0xc000cb0ea0, 0xc000681000)                                            
        /workspace/controllers/openstackcluster_controller.go:315 +0xbf6                                                                                                  
sigs.k8s.io/cluster-api-provider-openstack/controllers.reconcileNormal({0x2466e58, 0xc0011f5040}, 0x0?, 0xc000681000)                                                     
        /workspace/controllers/openstackcluster_controller.go:280 +0x125                                                                                                  
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackClusterReconciler).Reconcile(0xc000a983c0, {0x245ddb8, 0xc000638e10}, {{{0xc0009bf9b0?, 0x0?}, {0xc0003b
a8d0?, 0xc000d35d48?}}})                                                                                                                                                  
        /workspace/controllers/openstackcluster_controller.go:127 +0x5e6                                                                                                  
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2463430?, {0x245ddb8?, 0xc000638e10?}, {{{0xc0009bf9b0?, 0xb?}, {0xc0003ba8d0?, 0x0?}}}) 
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:119 +0xc8                                                                
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000168140, {0x245dd10, 0xc0002d2dc0}, {0x1ee04e0?, 0xc0008fe4c0?})               
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:316 +0x3ca                                                               
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000168140, {0x245dd10, 0xc0002d2dc0})                                         
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266 +0x1d9                                                               
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()                                                                                      
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227 +0x85                                                                
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2                                                                               
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:223 +0x587
```

--- a/api/v1alpha8/openstackcluster_types.go
+++ b/api/v1alpha8/openstackcluster_types.go
@@ -155,8 +155,8 @@ type OpenStackClusterSpec struct {
 	// Bastion is the OpenStack instance to login the nodes
 	//
 	// As a rolling update is not ideal during a bastion host session, we
-	// prevent changes to a running bastion configuration. Set `enabled: false` to
-	// make changes.
+	// prevent changes to a running bastion configuration. To make changes, it's required
+	// to first set `enabled: false` which will remove the bastion and then changes can be made.
 	//+optional
 	Bastion *Bastion `json:"bastion,omitempty"`
 

--- a/api/v1alpha8/openstackcluster_webhook.go
+++ b/api/v1alpha8/openstackcluster_webhook.go
@@ -116,6 +116,13 @@ func (r *OpenStackCluster) ValidateUpdate(oldRaw runtime.Object) (admission.Warn
 		r.Spec.APIServerPort = 0
 	}
 
+	// Allow to remove the bastion spec only if it was disabled before.
+	if r.Spec.Bastion == nil {
+		if old.Spec.Bastion != nil && old.Spec.Bastion.Enabled {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "bastion"), "cannot be removed before disabling it"))
+		}
+	}
+
 	// Allow changes to the bastion spec.
 	old.Spec.Bastion = &Bastion{}
 	r.Spec.Bastion = &Bastion{}

--- a/api/v1alpha8/openstackcluster_webhook_test.go
+++ b/api/v1alpha8/openstackcluster_webhook_test.go
@@ -358,6 +358,42 @@ func TestOpenStackCluster_ValidateUpdate(t *testing.T) {
 			},
 			wantErr: true,
 		},
+		{
+			name: "Removing OpenStackCluster.Spec.Bastion when it is enabled is not allowed",
+			oldTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					Bastion: &Bastion{
+						Enabled: true,
+						Instance: OpenStackMachineSpec{
+							Flavor: "m1.small",
+							Image:  ImageFilter{Name: "ubuntu"},
+						},
+					},
+				},
+			},
+			newTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{},
+			},
+			wantErr: true,
+		},
+		{
+			name: "Removing OpenStackCluster.Spec.Bastion when it is disabled is allowed",
+			oldTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					Bastion: &Bastion{
+						Enabled: false,
+						Instance: OpenStackMachineSpec{
+							Flavor: "m1.small",
+							Image:  ImageFilter{Name: "ubuntu"},
+						},
+					},
+				},
+			},
+			newTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{},
+			},
+			wantErr: false,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

--- a/config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml
+++ b/config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml
@@ -4899,8 +4899,8 @@ spec:
 
 
                   As a rolling update is not ideal during a bastion host session, we
-                  prevent changes to a running bastion configuration. Set `enabled: false` to
-                  make changes.
+                  prevent changes to a running bastion configuration. To make changes, it's required
+                  to first set `enabled: false` which will remove the bastion and then changes can be made.
                 properties:
                   availabilityZone:
                     type: string

--- a/config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclustertemplates.yaml
+++ b/config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclustertemplates.yaml
@@ -2324,8 +2324,8 @@ spec:
 
 
                           As a rolling update is not ideal during a bastion host session, we
-                          prevent changes to a running bastion configuration. Set `enabled: false` to
-                          make changes.
+                          prevent changes to a running bastion configuration. To make changes, it's required
+                          to first set `enabled: false` which will remove the bastion and then changes can be made.
                         properties:
                           availabilityZone:
                             type: string

--- a/controllers/openstackcluster_controller.go
+++ b/controllers/openstackcluster_controller.go
@@ -263,7 +263,10 @@ func deleteBastion(scope scope.Scope, cluster *clusterv1.Cluster, openStackClust
 			}
 		}
 
-		instanceSpec := bastionToInstanceSpec(openStackCluster, cluster.Name)
+		instanceSpec, err := bastionToInstanceSpec(openStackCluster, cluster.Name)
+		if err != nil {
+			return err
+		}
 		if err = computeService.DeleteInstance(openStackCluster, openStackCluster, instanceStatus, instanceSpec); err != nil {
 			handleUpdateOSCError(openStackCluster, fmt.Errorf("failed to delete bastion: %w", err))
 			return fmt.Errorf("failed to delete bastion: %w", err)
@@ -346,7 +349,10 @@ func reconcileBastion(scope scope.Scope, cluster *clusterv1.Cluster, openStackCl
 		return reconcile.Result{}, err
 	}
 
-	instanceSpec := bastionToInstanceSpec(openStackCluster, cluster.Name)
+	instanceSpec, err := bastionToInstanceSpec(openStackCluster, cluster.Name)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
 	bastionHash, err := compute.HashInstanceSpec(instanceSpec)
 	if err != nil {
 		return reconcile.Result{}, fmt.Errorf("failed computing bastion hash from instance spec: %w", err)
@@ -437,7 +443,15 @@ func reconcileBastion(scope scope.Scope, cluster *clusterv1.Cluster, openStackCl
 	return ctrl.Result{}, nil
 }
 
-func bastionToInstanceSpec(openStackCluster *infrav1.OpenStackCluster, clusterName string) *compute.InstanceSpec {
+func bastionToInstanceSpec(openStackCluster *infrav1.OpenStackCluster, clusterName string) (*compute.InstanceSpec, error) {
+	if openStackCluster.Spec.Bastion == nil {
+		return nil, fmt.Errorf("bastion spec is nil")
+	}
+
+	if openStackCluster.Status.Bastion == nil {
+		return nil, fmt.Errorf("bastion status is nil")
+	}
+
 	instanceSpec := &compute.InstanceSpec{
 		Name:          bastionName(clusterName),
 		Flavor:        openStackCluster.Spec.Bastion.Instance.Flavor,
@@ -458,7 +472,7 @@ func bastionToInstanceSpec(openStackCluster *infrav1.OpenStackCluster, clusterNa
 
 	instanceSpec.Ports = openStackCluster.Spec.Bastion.Instance.Ports
 
-	return instanceSpec
+	return instanceSpec, nil
 }
 
 func bastionName(clusterName string) string {

--- a/docs/book/src/clusteropenstack/configuration.md
+++ b/docs/book/src/clusteropenstack/configuration.md
@@ -35,6 +35,8 @@
   - [Custom pod network CIDR](#custom-pod-network-cidr)
   - [Accessing nodes through the bastion host via SSH](#accessing-nodes-through-the-bastion-host-via-ssh)
     - [Enabling the bastion host](#enabling-the-bastion-host)
+    - [Making changes to the bastion host](#making-changes-to-the-bastion-host)
+    - [Disabling the bastion](#disabling-the-bastion)
     - [Obtain floating IP address of the bastion node](#obtain-floating-ip-address-of-the-bastion-node)
 
 <!-- END doctoc generated TOC please keep comment here to allow auto update -->
@@ -653,6 +655,19 @@ spec:
 ```
 
 If `managedSecurityGroups: true`, security group rule opening 22/tcp is added to security groups for bastion, controller, and worker nodes respectively. Otherwise, you have to add `securityGroups` to the `bastion` in `OpenStackCluster` spec and `OpenStackMachineTemplate` spec template respectively.
+
+### Making changes to the bastion host
+
+Changes can be made to the bastion instance, like for example changing the flavor.
+First, you have to disable the bastion host by setting `enabled: false` in the `OpenStackCluster.Spec.Bastion` field.
+The bastion will be deleted, you can check the status of the bastion host by running `kubectl get openstackcluster` and looking at the `Bastion` field in status.
+Once it's gone, you can re-enable the bastion host by setting `enabled: true` and then making changes to the bastion instance spec by modifying the `OpenStackCluster.Spec.Bastion.Instance` field.
+The bastion host will be re-created with the new instance spec.
+
+### Disabling the bastion
+
+To disable the bastion host, set `enabled: false` in the `OpenStackCluster.Spec.Bastion` field. The bastion host will be deleted, you can check the status of the bastion host by running `kubectl get openstackcluster` and looking at the `Bastion` field in status.
+Once it's gone, you can now remove the `OpenStackCluster.Spec.Bastion` field from the `OpenStackCluster` spec.
 
 ### Obtain floating IP address of the bastion node
 

